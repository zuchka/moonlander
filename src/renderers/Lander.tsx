import React from 'react';
import { View, StyleSheet } from 'react-native';
import Matter from 'matter-js';
import Svg, { G, Polygon } from 'react-native-svg';

// Define props including the testIDs from the test file
interface LanderProps {
    body: Matter.Body;
    size: number[];
    testID?: string; // For the Svg container
    gTestID?: string; // For the G element
    polygonTestID?: string; // For the Polygon
}

const Lander: React.FC<LanderProps> = ({ body, size, testID, gTestID, polygonTestID }) => {
    const width = size[0];
    const height = size[1];
    const { x: bodyX, y: bodyY } = body.position;
    const angleRad = body.angle;
    const angleDeg = angleRad * 180 / Math.PI; // Convert to degrees for SVG transform

    // Calculate top-left position for the Svg container
    const topLeftX = bodyX - width / 2;
    const topLeftY = bodyY - height / 2;

    // Define the shape of the lander relative to its center (0,0 within the G element)
    // Simple triangular shape for example:
    const points = [
        `${0},${-height / 2}`,         // Top point
        `${-width / 2},${height / 2}`, // Bottom left
        `${width / 2},${height / 2}`   // Bottom right
    ].join(' ');

    return (
        <Svg
            testID={testID}
            width={width}
            height={height}
            style={[
                styles.landerSvg,
                {
                    left: topLeftX,
                    top: topLeftY,
                }
            ]}
            // viewBox might be useful if shape coordinates are complex
            // viewBox={`0 0 ${width} ${height}`}
        >
            <G
                testID={gTestID} // Pass down testID
                rotation={angleDeg}
                originX={width / 2} // Rotate around the center of the SVG container
                originY={height / 2}
                // Apply translation to center the shape within the G element before rotation
                translateX={width / 2}
                translateY={height / 2}
            >
                <Polygon
                    testID={polygonTestID} // Pass down testID
                    points={points}
                    fill="lightblue" // Example color
                    stroke="white"
                    strokeWidth="1"
                />
                {/* Add thruster elements here later */}
            </G>
        </Svg>
    );
};

const styles = StyleSheet.create({
    landerSvg: {
        position: 'absolute',
        // backgroundColor: 'rgba(255, 0, 0, 0.1)', // Optional: for debugging bounds
    },
});

export default Lander; 