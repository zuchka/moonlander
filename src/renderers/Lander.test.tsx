import React from 'react';
import { render } from '@testing-library/react-native';
import Lander from './Lander';
import Matter from 'matter-js';
import Svg, { G, Polygon } from 'react-native-svg';

// Mock Body
const mockBody: Matter.Body = {
    position: { x: 150, y: 250 },
    angle: Math.PI / 4, // 45 degrees
} as Matter.Body;

const size = [40, 40]; // width, height

describe('<Lander /> Renderer', () => {

    test('renders an Svg container with correct size', () => { // Renamed test slightly
        const { getByTestId } = render(<Lander body={mockBody} size={size} testID="lander-svg" />);
        const svg = getByTestId('lander-svg');

        expect(svg).toBeDefined();
        // Check basic width/height props, removed complex style check
        expect(svg.props.width).toBe(size[0]);
        expect(svg.props.height).toBe(size[1]);
    });

    test('renders an inner G element', () => { // Simplified test name further
        // Render with a testID on the G element
        const { getByTestId } = render(
            <Lander body={mockBody} size={size} testID="lander-svg" gTestID="lander-g" />
        );
        const g = getByTestId('lander-g');

        expect(g).toBeDefined();
        // Removed checks for g.props.rotation, originX, originY
    });

    test('renders a Polygon element inside the G', () => {
        const { getByTestId } = render(
            <Lander body={mockBody} size={size} testID="lander-svg" polygonTestID="lander-polygon" />
         );
        const polygon = getByTestId('lander-polygon');

        expect(polygon).toBeDefined();
        // Removed check for polygon.props.points
        expect(polygon.props.fill).toBeDefined(); // Keep basic style check
    });

    // Optional: Snapshot test to catch unexpected structure changes
    test('matches snapshot', () => {
        const tree = render(<Lander body={mockBody} size={size} />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});

// Helper type to allow passing testIDs down for testing internals
type LanderProps = {
  body: Matter.Body;
  size: number[];
  testID?: string;
  gTestID?: string;
  polygonTestID?: string;
};