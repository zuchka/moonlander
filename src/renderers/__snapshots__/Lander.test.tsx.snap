// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Lander /> Renderer matches snapshot 1`] = `
<RNSVGSvgView
  bbHeight={40}
  bbWidth={40}
  focusable={false}
  height={40}
  style={
    [
      {
        "backgroundColor": "transparent",
        "borderWidth": 0,
      },
      [
        {
          "position": "absolute",
        },
        {
          "left": 130,
          "top": 230,
        },
      ],
      {
        "flex": 0,
        "height": 40,
        "width": 40,
      },
    ]
  }
  width={40}
>
  <RNSVGGroup
    fill={
      {
        "payload": 4278190080,
        "type": 0,
      }
    }
  >
    <RNSVGGroup
      fill={
        {
          "payload": 4278190080,
          "type": 0,
        }
      }
      matrix={
        [
          0.7071067811865476,
          0.7071067811865475,
          -0.7071067811865475,
          0.7071067811865476,
          40,
          11.715728752538098,
        ]
      }
    >
      <RNSVGPath
        d="M0 -20 -20 20 20 20z"
        fill={
          {
            "payload": 4289583334,
            "type": 0,
          }
        }
        propList={
          [
            "fill",
            "stroke",
            "strokeWidth",
          ]
        }
        stroke={
          {
            "payload": 4294967295,
            "type": 0,
          }
        }
        strokeWidth="1"
      />
    </RNSVGGroup>
  </RNSVGGroup>
</RNSVGSvgView>
`;
